orm/primary_key.go

```diff
@@ -30,10 +32,54 @@ type PrimaryKeyed interface {
 	//
 	// The `IndexKeyCodec` used with the `PrimaryKeyTable` may add certain constraints to the byte representation as
 	// max length = 255 in `Max255DynamicLengthIndexKeyCodec` or a fix length in `FixLengthIndexKeyCodec` for example.
-	PrimaryKey() []byte
+	// PrimaryKey() []byte
+	PrimaryKeyFields() []interface{}
 	codec.ProtoMarshaler
 }
 
+func PrimaryKey(obj PrimaryKeyed) []byte {
+	fields := obj.PrimaryKeyFields()
+	return buildPrimaryKey(fields...)
+}
+
+// This can be strings, bytes, or some integer stuff
+// Strings will be null-terminated, bytes will be length-prefixed, ...
+func buildPrimaryKey(fields ...interface{}) []byte {
+	bytesSlice := make([][]byte, len(fields))
+	totalLen := 0
+	for i, field := range fields {
+		bytesSlice[i] = primaryKeyFieldBytes(field)
+		totalLen += len(bytesSlice[i])
+	}
+	primaryKey := make([]byte, 0, totalLen)
+	for _, bs := range bytesSlice {
+		primaryKey = append(primaryKey, bs...)
+	}
+	return primaryKey
+}
+
+func primaryKeyFieldBytes(field interface{}) []byte {
+	switch v := field.(type) {
+	case []byte:
+		return AddLengthPrefix(v)
+	case string:
+		// TODO(@ruhatch): Do something with strings
+		return []byte(v)
```

Comment:
We have `UnsafeBytesToString` which we should use here to remove unnecessary memory allocations

Reply:
BTW: why we don't Len prefix strings?

Reply:
In the discussion [here](https://github.com/cosmos/cosmos-sdk/discussions/9156), @aaronc suggested using null termination on strings. I haven't implemented it yet, but was planning on doing so if the general approach looked okay.